state-chain/pallets/cf-ingress-egress/src/lib.rs

```diff
@@ -530,35 +557,70 @@ pub mod pallet {
 	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
 		/// Recycle addresses if we can
 		fn on_idle(_n: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
-			let read_write_weight =
-				frame_support::weights::constants::RocksDbWeight::get().reads_writes(1, 1);
-
-			let maximum_recycle_number = remaining_weight
-				.ref_time()
-				.checked_div(read_write_weight.ref_time())
-				.unwrap_or_default()
-				.saturated_into::<usize>();
-
-			let can_recycle = DepositChannelRecycleBlocks::<T, I>::mutate(|recycle_queue| {
-				Self::can_and_cannot_recycle(
-					recycle_queue,
-					maximum_recycle_number,
-					T::ChainTracking::get_block_height(),
-				)
-			});
+			let mut used_weight = Weight::zero();
+
+			let read_weight =
+				frame_support::weights::constants::RocksDbWeight::get().reads_writes(1, 0);
+
+			while remaining_weight.saturating_sub(used_weight).all_gte(read_weight * 2) {
+				if let Some((block, address)) =
+					DepositChannelRecycleBlocks::<T, I>::get().iter().next()
+				{
+					used_weight = used_weight.saturating_add(read_weight);
 
-			for address in can_recycle.iter() {
-				if let Some(details) = DepositChannelLookup::<T, I>::take(address) {
-					if let Some(state) = details.deposit_channel.state.maybe_recycle() {
-						DepositChannelPool::<T, I>::insert(
+					if *block > T::ChainTracking::get_block_height() {
+						break
+					}
+
+					used_weight = used_weight.saturating_add(read_weight);
+
+					if let Some(details) = DepositChannelLookup::<T, I>::take(address) {
+						let item_count = BoostableDeposits::<T, I>::iter_prefix(
 							details.deposit_channel.channel_id,
-							DepositChannel { state, ..details.deposit_channel },
+						)
+						.count() as u32;
+						let _removed_deposits = BoostableDeposits::<T, I>::clear_prefix(
+							details.deposit_channel.channel_id,
+							item_count,
+							None,
 						);
+
+						let cleanup_weight = T::WeightInfo::remove_boostable_deposits(item_count)
+							.saturating_add(
+								frame_support::weights::constants::RocksDbWeight::get()
+									.reads_writes(0, 2),
+							);
+
+						if !remaining_weight.saturating_sub(used_weight).all_gte(cleanup_weight) {
```

Comment:
this is kinda dangerous no? We've taken a channel from the storage. If this happens, we won't recycle the address, and nor will we on the next iteration.

Reply:
yeah i was not finished refactoring the code. I want to do only reads to calculate the exact weight before starting the actual cleanup.